libraries/chain/controller.cpp

```diff
@@ -1018,14 +1018,20 @@ struct controller_impl {
 
 
    void push_block( const signed_block_ptr& b, controller::block_status s ) {
-    //  idump((fc::json::to_pretty_string(*b)));
       EOS_ASSERT(!pending, block_validate_exception, "it is not valid to push a block when there is a pending block");
+
+      auto reset_prod_light_validation = fc::make_scoped_exit([this]() {
+         conf.trusted_producer_light_validation = false;
+      });
       try {
          EOS_ASSERT( b, block_validate_exception, "trying to push empty block" );
          EOS_ASSERT( s != controller::block_status::incomplete, block_validate_exception, "invalid block status for a completed block" );
          emit( self.pre_accepted_block, b );
          bool trust = !conf.force_all_checks && (s == controller::block_status::irreversible || s == controller::block_status::validated);
          auto new_header_state = fork_db.add( b, trust );
+         if (conf.trusted_producers.count(b->producer)) {
+            conf.trusted_producer_light_validation = true;
+         };
```

Comment:
Instead of setting a flag, I wonder if we could just add this check in `light_validation_allowed` then the extra flag and reset would not be needed.

Reply:
I think that would be much more complicated (handling identifying if we are speculating or applying a block) and this gets called multiple times in a transaction pass, versus being called once for the block.